File: isegoria/src/main/java/com/eulersbridge/isegoria/network/UnwrapConverterFactory.java
package com.eulersbridge.isegoria.network;

import android.support.annotation.Nullable;

import com.eulersbridge.isegoria.network.api.responses.GenericPaginatedResponse;

import java.lang.annotation.Annotation;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

import okhttp3.ResponseBody;
import retrofit2.Converter;
import retrofit2.Retrofit;
import retrofit2.converter.moshi.MoshiConverterFactory;

/**
 * "Unwraps" API methods annotated with @Paginated by returning the contents of their `foundObjects`
 * rather than the entire response.
 */
class UnwrapConverterFactory extends Converter.Factory {

    private final MoshiConverterFactory factory;

    UnwrapConverterFactory(MoshiConverterFactory factory) {
        this.factory = factory;
    }

    @Override
    public @Nullable
    Converter<ResponseBody, ?> responseBodyConverter(Type type,
                                                     Annotation[] annotations, Retrofit retrofit) {

        for (Annotation annotation : annotations) {
            if (annotation.annotationType() == Paginated.class) {

                Type wrappedType = new ParameterizedType() {
                    @Override
                    public Type[] getActualTypeArguments() {
                        return new Type[] {type};
                    }

                    @Override
                    public Type getOwnerType() {
                        return null;
                    }

                    @Override
                    public Type getRawType() {
                        return GenericPaginatedResponse.class;
                    }
                };

                Converter<ResponseBody, ?> converter = factory
                        .responseBodyConverter(wrappedType, annotations, retrofit);

                //noinspection unchecked,unchecked
                return new ResponseWrapperBodyConverter(converter);
            }
        }

        return null;
    }

}

File: isegoria/src/main/kotlin/com/eulersbridge/isegoria/network/UnwrapConverterFactory.kt
package com.eulersbridge.isegoria.network

import com.eulersbridge.isegoria.network.api.Paginated
import com.eulersbridge.isegoria.network.api.response.GenericPaginatedResponse
import com.squareup.moshi.Types
import okhttp3.ResponseBody
import retrofit2.Converter
import retrofit2.Retrofit
import java.lang.reflect.Type

/**
 * "Unwraps" API methods annotated with @Paginated by returning the contents of their `foundObjects`
 * rather than the entire response.
 */
internal class UnwrapConverterFactory : Converter.Factory() {

    override fun responseBodyConverter(
        type: Type,
        annotations: Array<Annotation>, retrofit: Retrofit
    ): Converter<ResponseBody, *>? {

        return if (annotations.any { it.annotationClass == Paginated::class }) {

            val envelopeType = Types.newParameterizedType(GenericPaginatedResponse::class.java, type)

            val delegate = retrofit.nextResponseBodyConverter<GenericPaginatedResponse<*>>(this, envelopeType, annotations)

            Converter<ResponseBody, Any> { value ->
                val (foundObjects) = delegate.convert(value)

                foundObjects
                // NOTE: there was previously a check to return null if found objects is an empty
                //       list, this was removed because it was causing unnecessary errors and had no
                //        obvious purpose
            }

        } else {
            null
        }
    }

}
